### PR TITLE
8289230: Move PlatformXXX class declarations out of os_xxx.hpp

### PR DESCRIPTION
**Pull request by iklam**
There are only two implementations of these classes (one for windows, and one for posix):

- PlatformEvent
- PlatformParker
- PlatformMutex
- PlatformMonitor
- ThreadCrashProtection

Before this PR, these classes are declared in os_xxx.hpp. This causes excessive inclusion of the large header file os.hpp by popular headers such as mutex.hpp, which needs only the declaration of PlatformMutex but not the other stuff in os.hpp

This PR moves the declarations to park_posix.hpp, mutex_posix.hpp, etc.

Note:  ideally, the definition of PlatformParker/PlatformEvent should be moved to park_posix.cpp, and PlatformMutex/PlatformMonitor should be moved to mutex_posix.cpp. However, the definition of these 4 classes are intertwined, so I'll leave them inside os_posix.cpp for now. (Same for the Windows version).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289230](https://bugs.openjdk.org/browse/JDK-8289230): Move PlatformXXX class declarations out of os_xxx.hpp

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [8ceda0f5](https://git.openjdk.org/jdk/pull/9303/files/8ceda0f5c2ec1f5b5536a6806ec1090cac358a4f)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [9d502ffb](https://git.openjdk.org/jdk/pull/9303/files/9d502ffbeda622cc625211af639b1e4a42fbfeb0)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9303/head:pull/9303` \
`$ git checkout pull/9303`

Update a local copy of the PR: \
`$ git checkout pull/9303` \
`$ git pull https://git.openjdk.org/jdk pull/9303/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9303`

View PR using the GUI difftool: \
`$ git pr show -t 9303`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9303.diff">https://git.openjdk.org/jdk/pull/9303.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/os/posix/mutex_posix.hpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+ private:
+  NONCOPYABLE(PlatformMonitor);
+
+ public:
+  int wait(jlong millis);
+  void notify();
+  void notify_all();
+};
+
+#endif // OS_POSIX_PARK_POSIX_HPP
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/mutex_posix.hpp:141]: Wrong comment.

File: src/hotspot/os/posix/threadCrashProtection_posix.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+Thread* ThreadCrashProtection::_protected_thread = NULL;
+ThreadCrashProtection* ThreadCrashProtection::_crash_protection = NULL;
+
+ThreadCrashProtection::ThreadCrashProtection() {
+  _protected_thread = Thread::current();
+  assert(_protected_thread->is_JfrSampler_thread(), "should be JFRSampler");
+}
+
+/*
+ * See the caveats for this class in os_posix.hpp
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/threadCrashProtection_posix.cpp:38]: Comment needs updating

File: src/hotspot/os/windows/mutex_windows.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+ public:
+  PlatformMonitor();
+  ~PlatformMonitor();
+  int wait(jlong millis);
+  void notify();
+  void notify_all();
+};
+
+#endif // OS_WINDOWS_PARK_WINDOWS_HPP
```
Comment (dholmes-ora) [inline src/hotspot/os/windows/mutex_windows.hpp:64]: Wrong comment

**Review by dholmes-ora — COMMENTED** [review]
Just took a quick skim through to get the general sense of things. Header file split is okay. Pity about the .cpp situation though - maybe move to platform_posix.cpp and platform_windows.cpp to at least get them out of the os_xxx.cpp file?

File: src/hotspot/os/posix/mutex_posix.hpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/os/posix/mutex_posix.hpp]: Fixed.

File: src/hotspot/share/runtime/mutex.hpp [hunk 4]
```diff
@@ -192,7 +200,7 @@ class Mutex : public CHeapObj<mtSynchronizer> {
   void print_on_error(outputStream* st) const;
   #ifndef PRODUCT
     void print_on(outputStream* st) const;
-    void print() const                      { print_on(::tty); }
+    void print() const                      { /*print_on(::tty); */ } // FIXME
```
Comment (coleenp) [inline src/hotspot/share/runtime/mutex.hpp:203]: Can you move this print implementation into the .cpp file?

File: src/hotspot/share/runtime/threadCrashProtection.hpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  virtual void call() = 0;
+};
+
+#if defined(LINUX) || defined(AIX) || defined(BSD)
+# include "threadCrashProtection_posix.hpp"
+#elif defined(_WINDOWS)
+# include "threadCrashProtection_windows.hpp"
+#else
+# error "No ThreadCrashProtection implementation provided for this OS"
+#endif
```
Comment (coleenp) [inline src/hotspot/share/runtime/threadCrashProtection.hpp:42]: Shouldn't you use this?
#define OS_HEADER(basename)            XSTR(OS_HEADER_STEM(basename).hpp)

**Review by coleenp — APPROVED** [review]
I very much like this incremental approach to reducing our over-inclusion.

File: src/hotspot/share/runtime/threadCrashProtection.hpp (hunk 5 above)
Comment (iklam) [reply src/hotspot/share/runtime/threadCrashProtection.hpp:40]: I fixed it as you suggested. I also fixed semaphore.hpp where I copied the old pattern from.

File: src/hotspot/share/runtime/mutex.hpp (hunk 4 above)
Comment (iklam) [reply src/hotspot/share/runtime/mutex.hpp]: Fixed.

File: src/hotspot/share/runtime/threadCrashProtection.hpp (hunk 5 above)
Comment (coleenp) [reply src/hotspot/share/runtime/threadCrashProtection.hpp:40]: Doesn't the OS_HEADER() macro handle the posix case too?

Comment (iklam) [reply src/hotspot/share/runtime/threadCrashProtection.hpp:40]: ` OS_HEADER(threadCrashProtection)` will give threadCrashProtection_linux.hpp, etc. It won't give us threadCrashProtection_posix.hpp. I didn't want to write three threadCrashProtection_{aix,bsd,linux}.hpp files that each just has a single line that includes the _posix version.

Comment (dholmes-ora) [reply src/hotspot/share/runtime/threadCrashProtection.hpp:40]: Yeah no posix support with OS_HEADER.

File: src/hotspot/share/runtime/mutex.cpp [hunk 6]
```diff
@@ -331,7 +332,6 @@ const char* Mutex::rank_name() const {
   return rank_name_internal(_rank);
 }
 
-
```
Comment (calvinccheung) [inline src/hotspot/share/runtime/mutex.cpp:334]: Line 334 deleted by accident?

**Review by calvinccheung — APPROVED** [review]
Looks good. Just one nit.

**Comment by iklam** [comment]
Thanks @coleenp @calvinccheung @dholmes-ora for the review.
/integrate

**Review by kristylee88 — APPROVED** [review]

**Review by kristylee88 — APPROVED** [review]

**Pull request closed without merging**